### PR TITLE
Feature/sof 6006

### PR DESCRIPTION
**Pull request by azech-hqs**

### CODE REVIEW

File: src/js/element.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
+import { ELEMENT_COLORS } from "./element_colors";
+import { PERIODIC_TABLE } from "./periodic_table.js";
+import { CONVERSION, UNITS, convertLength, convertEnergy } from "./units";
+
+export class Element {
```
Comment (timurbazhirov) [inline src/js/element.js]: Let's use `ChemicaElement` pls, for clarity. In JS there are many other Elements - HTML, https://element.eleme.io/#/en-US, etc.

File: src/js/element.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    constructor(symbol) {
```
Comment (timurbazhirov) [inline src/js/element.js:10]: We should add a docstring that explains the parameter type for `symbol`

File: src/js/element.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        this.symbol = symbol;
```
Comment (timurbazhirov) [inline src/js/element.js:11]: based on the getter/setter logic below this line should be:
```js
this._symbol = symbol
```

Comment (tjduigna) [reply src/js/element.js:11]: Assigning `this._symbol` directly explicitly avoids the logic in the setter whereas `this.symbol = symbol` forces the use of the setter. This is the correct usage.

File: src/js/element.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    get ionizationPotential() {
+        if (Number.isFinite(this._properties.first_ionizing_kJ_mol)) {
+            return (
+                this._properties.first_ionizing_kJ_mol /
+                CONVERSION.energy[UNITS.energy.electronvolt][UNITS.energy.kJ_mol]
+            );
+        }
+    }
+
+    ionizationPotentialIn(unit) {
```
Comment (timurbazhirov) [inline src/js/element.js]: `getIonizationPotentialInUnits`

File: src/js/element.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            throw new Error(`Invalid symbol: ${symbol}`);
+        }
+    }
+
+    /**
+     * @summary Returns atomic mass in amu.
+     * @returns {number|undefined} - Atomic mass in amu
+     */
+    get mass() {
+        if (Number.isFinite(this._properties.atomic_mass)) {
```
Comment (timurbazhirov) [inline src/js/element.js]: 1. The pattern is a bit repetitive - maybe we should define a checker function
2. `typeof` would probably be a more clear check

File: src/js/element.js (hunk 4 above)
Comment (timurbazhirov) [reply src/js/element.js]: On a second thought, I see your intent. Maybe let's call the unit functions `<property>InUnits`

File: src/js/index.js [hunk 6]
```diff
@@ -101,11 +90,34 @@ export function getElementsBondsData(element1, element2, order = undefined) {
...
     return 1.0;
 }
+
+/**
+ *
+ * @param elements {String[]} - List of elements (as symbols)
+ * @param properties {String[]} - List of properties
+ * @param propMap {Object} - Object mapping property names to custom names (e.g. {'atomic_radius_pm': 'radius'})
+ * @param separator {String} - Separator string for the suffix (default: ':')
+ * @returns {Object[]} - List of Objects {property_key: property_value}
```
Comment (timurbazhirov) [inline src/js/index.js:109]: It would probably be good to provide an example of what this returns, for clarity.

Comment (timurbazhirov) [reply src/js/index.js:109]: See https://github.com/Exabyte-io/express/blob/a182f7dccfc5c715aa4ca4adf0c20be85f90f7b7/express/parsers/mixins/ionic.py for reference

File: src/js/units.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    },
+    mass: {
+        kilogram: "kg",
+        dalton: "Da",
+        atomicMassUnit: "amu",
+    },
+};
+
+/**
+ * Conversion coefficients based on data taken from Wikipedia (date accessed: 05/20/2022)
```
Comment (timurbazhirov) [inline src/js/units.js:27]: This would need a better set of references ...

File: src/js/units.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    if (from === to) return value;
+
+    if (CONVERSION.energy.hasOwnProperty(from) && CONVERSION.energy[from].hasOwnProperty(to)) {
+        return value * CONVERSION.energy[from][to];
+    } else if (CONVERSION.energy.hasOwnProperty(to) && CONVERSION.energy[to].hasOwnProperty(from)) {
+        return value / CONVERSION.energy[to][from];
+    } else {
+        console.warn([value, from, to], "Unable to convert unit of energy!");
+        return undefined;
+    }
```
Comment (timurbazhirov) [inline src/js/units.js:110]: The two functions look pretty much the same above. We can abstract/parametrize probably

File: tests/js/index.js [hunk 9]
```diff
@@ -3,34 +3,36 @@ import { expect } from "chai";
 import {
     defaultElementsBondsDataEntry,
     ELEMENT_COLORS,
-    ELEMENT_VDW_RADII,
-    ELEMENTS_BY_SYMBOL,
     filterBondsDataByElementsAndOrder,
     getElectronegativity,
     getElementsBondsData,
     PERIODIC_TABLE,
+    Element,
```
Comment (timurbazhirov) [inline tests/js/index.js]: We should consider adding tests for the new functionality

File: src/js/element.js (hunk 3 above)
Comment (timurbazhirov) [reply src/js/element.js:11]: OK

File: src/js/element.js (hunk 1 above)
Comment (azech-hqs) [reply src/js/element.js]: Class name has been renamed

File: src/js/element.js (hunk 4 above)
Comment (azech-hqs) [reply src/js/element.js]: Function has been renamed.

File: src/js/element.js (hunk 5 above)
Comment (azech-hqs) [reply src/js/element.js]: I used `typeof` instead. For the checker function, perhaps we can validate using schemas here?

File: src/js/index.js (hunk 6 above)
Comment (azech-hqs) [reply src/js/index.js:109]: Example input and output were added to the docstring.

File: src/js/units.js (hunk 7 above)
Comment (azech-hqs) [reply src/js/units.js:27]: I added the links to the corresponding Wikipedia pages for reference.

File: src/js/units.js (hunk 8 above)
Comment (azech-hqs) [reply src/js/units.js:110]: The functions have been refactored to a single function `convertUnit`.

File: tests/js/index.js (hunk 9 above)
Comment (azech-hqs) [reply tests/js/index.js]: New tests were added for `getAtomicPropertiesFlat()` and functions of `ChemicalElement`.

File: src/js/index.js [hunk 10]
```diff
@@ -101,11 +92,54 @@ export function getElementsBondsData(element1, element2, order = undefined) {
...
+ * //  { 'atomic_number:Na': 11 },
+ * //  { 'pauling_negativity:H': 2.2 },
+ * //  { 'pauling_negativity:Na': 0.93 },
+ * // ]
+ */
+export function getAtomicPropertiesFlat(
+    elements,
+    properties,
+    propertiesMap = undefined,
+    separator = ":",
```
Comment (timurbazhirov) [inline src/js/index.js:130]: Let's make this to be an object with named keys for clarity, pls

File: src/js/index.js [hunk 11]
```diff
@@ -101,11 +92,54 @@ export function getElementsBondsData(element1, element2, order = undefined) {
...
+ * );
+ * // returns [
+ * //  { 'atomic_radius:H': 25 },
+ * //  { 'atomic_radius:Na': 180 },
+ * //  { 'atomic_number:H': 1 },
+ * //  { 'atomic_number:Na': 11 },
+ * //  { 'pauling_negativity:H': 2.2 },
+ * //  { 'pauling_negativity:Na': 0.93 },
+ * // ]
+ */
```
Comment (timurbazhirov) [inline src/js/index.js:125]: We can remove `//` in the example

Comment (azech-hqs) [reply src/js/index.js:125]: Unnecessary comments have been removed and on my end, it still seems to render correctly.

**Review by timurbazhirov — APPROVED** [review]

**Pull request merged**